lib/core/src/Cardano/Wallet/DB/Sqlite/AddressBook.hs

```diff
@@ -0,0 +1,228 @@
+    AddressBook isomorphism
+-------------------------------------------------------------------------------}
+-- | FIXME LATER during ADP-1043:
+-- Move 'Prologue' and 'Discoveries' closer into address types.
+class AddressBookIso s where
+    -- | Address information contained in the prologue of the address book,
+    -- such as public keys or the address gap.
+    data Prologue s :: Type
+    -- | Addresses that were collected during discovery on the blockchain.
+    data Discoveries s :: Type
+
+    -- | Isomorphism between the address book type 's'
+    -- and its two components.
+    addressIso :: Iso' s (Prologue s, Discoveries s)
+
+{-------------------------------------------------------------------------------
+    Sequential address book
+-------------------------------------------------------------------------------}
+-- | Sequential list of addresses
+newtype SeqAddressList (c :: Role) = SeqAddressList [(W.Address, W.AddressState)]
+
+-- piggy-back on SeqState existing instance, to simulate the same behavior.
+instance AddressBookIso (Seq.SeqState n k)
+    => AddressBookIso (Seq.SeqAnyState n k p)
+  where
+    data Prologue (Seq.SeqAnyState n k p) = PS (Prologue (Seq.SeqState n k))
+    data Discoveries (Seq.SeqAnyState n k p) = DS (Discoveries (Seq.SeqState n k))
+
+    addressIso = withIso addressIso $ \from to ->
+        let from2 st = let (a,b) = from $ Seq.innerState st in (PS a, DS b)
+            to2 (PS a, DS b) = Seq.SeqAnyState $ to (a,b)
+        in  iso from2 to2
+
+-- | Isomorphism for sequential address book.
+instance 
+    ( MkKeyFingerprint key (Proxy n, key 'AddressK XPub)
+    , GetPurpose key
+    , SoftDerivation key
+    , PaymentAddress n key
+    , Typeable n
+    , (key == SharedKey) ~ 'False
+    ) => AddressBookIso (Seq.SeqState n key)
+  where
+    data Prologue (Seq.SeqState n key)
+        = SeqPrologue (Seq.SeqState n key)
+        -- Trick: We keep the type, but we empty the discovered addresses
+    data Discoveries (Seq.SeqState n key)
+        = SeqDiscoveries
+            (SeqAddressList 'UtxoInternal)
+            (SeqAddressList 'UtxoExternal)
+
+    addressIso = iso from to
+      where
+        from (Seq.SeqState int ext a b c) =
+            ( SeqPrologue $ Seq.SeqState (emptyPool int) (emptyPool ext) a b c
+            , SeqDiscoveries (toDiscoveries @n int) (toDiscoveries @n ext)
+            )
+        to (SeqPrologue (Seq.SeqState int ext a b c), SeqDiscoveries ints exts)
+          = Seq.SeqState (fromDiscoveries @n int ints) (fromDiscoveries @n ext exts) a b c
+
+-- | Extract the discovered addresses from an address pool.
+toDiscoveries
+    :: forall (n :: NetworkDiscriminant) (c :: Role) key.
+    ( GetPurpose key
+    , PaymentAddress n key
+    , Typeable c
+    ) => Seq.AddressPool c key -> SeqAddressList c
+toDiscoveries pool = SeqAddressList
+    [ (a,st) | (a,st,_) <- Seq.addresses (liftPaymentAddress @n) pool ]
```

Comment:
We do like this because SeqAddressList is directly mapped to DB, right?

Reply:
if yes, then for the future reference we need to remember about some probable future developments here:
1. multi accountsupport - in that case liftPaymentAddress will not be enough as payment credential is not enough. We would need to also differentiate against staking credential so staking key hash (for Sequential mode) or staking script hash (for Shared mode)
2. we have (at least in Sequential mode) several fingerprints of address to support: the current one (aka base with key hash in payment credential), enterprise address (key hash payment credenital without staking credential) and pointer address (key hash payment with pointer stake credential)
We need (probably later) to discuss if new abstraction are able to seamlessly support them

Reply:
> We do like this because SeqAddressList is directly mapped to DB, right?

Yes, indeed. Will be changed again in the next steps.

> […] staking credential […]

That's a very good point. From looking at the code, I got the impression that the Sequential state only looks at the payment key hashes. In other words, whenever the state sees an address on chain (which may have a stake key hash), it will extract the payment key hash, add it to the address pool as appropriate, convert it to an enterprise address and store that address in the database. Put differently, the database stores discovered Sequential payment key hashes, but formatted as enterprise addresses.

The situation for Shared states was not entirely clear to me, but by chasing the various type classes, I also got the impression that only the payment key hash is subject to discovery. In other words, if an address appears on the chain, and the stake credential of this address does not match the delegation template, but the payment credential does match the payment template, then it will still be discovered and added to the database. Makes sense: If you own the payment credential, you can always reassign the delegation part by making a transaction; the access control on the payment credential partially overrides the access control on the delegation part. I suppose we will need to think about this corner case again. 🤔 

In any case, for the time being, this PR only attempts to do what the code did before, no changes in functionality intended. 😅

Reply:
so for shared wallets we store payment credential only which is script hash. shared wallet's credential and sequential wallet credential happen to have the same byte imprint as the same hashes (blake224) in both cases is used.

What is important here is that maybe to avoid complications and choose simplicity it would be better to store both credentials, whatever they are represented (keyhash, scripthash or three numbers for pointer addresses). It would allow support of any address type and also seamlessly support multiaccount (on the level of discovery) 

I feel we will need to revisit this soon or later (maybe even during shared wallet work soon). I also see challenges (db migration issues....)

Reply:
> I feel we will need to revisit this soon or later (maybe even during shared wallet work soon). I also see challenges (db migration issues....)

Agreed. I suggest that we leave things as they are for now and revisit this soon during the shared wallet work.